dial_sync.go

```diff
@@ -53,22 +54,38 @@ func (ad *activeDial) incref() {
 
 func (ad *activeDial) decref() {
 	ad.refCntLk.Lock()
-	defer ad.refCntLk.Unlock()
 	ad.refCnt--
-	if ad.refCnt <= 0 {
-		ad.cancel()
+	maybeZero := (ad.refCnt <= 0)
+	ad.refCntLk.Unlock()
+
+	// make sure to always take locks in correct order.
+	if maybeZero {
 		ad.ds.dialsLk.Lock()
-		delete(ad.ds.dials, ad.id)
+		ad.refCntLk.Lock()
+		// check again after lock swap drop to make sure nobody else called incref
+		// in between locks
+		if ad.refCnt <= 0 {
+			ad.cancel()
+			delete(ad.ds.dials, ad.id)
+		}
 		ad.ds.dialsLk.Unlock()
+		ad.refCntLk.Unlock()
 	}
```

Comment:
Nit: I would release ad.refCntLk.Unlock() first, don't think it makes any difference in this case.